src/persist-client/src/cache.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A cache of [PersistClient]s indexed by [PersistLocation]s.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use mz_ore::metrics::MetricsRegistry;
+use mz_persist::cfg::{BlobMultiConfig, ConsensusConfig};
+use mz_persist::location::{BlobMulti, Consensus, ExternalError};
+use tokio::sync::Mutex;
+use tracing::debug;
+
+use crate::{PersistClient, PersistLocation};
+
+/// A cache of [PersistClient]s indexed by [PersistLocation]s.
+///
+/// There should be at most of these per process. All production PersistClients
+/// should be created through this cache.
+///
+/// This is because, in production, persist is heavily limited by the number of
+/// server-side Postgres/Aurora connections. This cache allows PersistClients to
+/// share, for example, these Postgres connections.
+#[derive(Debug, Clone)]
+pub struct PersistClientCache {
+    blob_by_uri: Arc<Mutex<HashMap<String, Arc<dyn BlobMulti + Send + Sync>>>>,
```

Comment:
I don't think this is the right place to put `Arc<Mutex<T>>` because this cache could still be useful to a worker that keeps it in a local variable and threads it through function calls and if someone wants to have a process global thing they can make an `Arc<Mutex<PersistClientCache>>` and store it in a static

Reply:
hmmm I think I've got to push back on this. the point of the cache is to minimize pg connections, which will be a constraint in production, so there should always be at most one of these per-process (though I'd imagine it as some field on the Server instance, not a static). plus the cost of the Arc<Mutex<>> is a rounding error compared to what happens on a cache miss

do you have a known need for a worker-local cache or is that hypothetical?

Reply:
okay, I'll pull the locking out. I'm still curious why you're talking about using this worker-local though? you understand that there needs to be at most one of these per process, right? is there a situation where we're guaranteed to have one worker per process?

Reply:
Mutex is gone

Reply:
sorry, I was responding to this:

> though I'd imagine it as some field on the Server instance, not a static

I'll add an `Arc<Mutex<PersistClientCache>>` field to each worker-local process instead of having a static